### ISSUE TITLE
`where Self: Sized` on trait method doesn't work as expected when implementing for unsized types.

### ISSUE THREAD

**Issue by crlf0710** [rust-lang/rust#57900]
I've got this code
````
trait Foo {
    fn bar(&self) -> Self where Self: Sized;
}

impl Foo for str {
//    fn bar(&self) -> Self { unimplemented!() }
}
````
the error is:
````
error[E0046]: not all trait items implemented, missing: `bar`
 --> src/lib.rs:5:1
````

However, since `str` is not `Sized`, i don't think `bar`  needs to be implemented. if i remove the "//", i'll got:
````
error[E0277]: the size for values of type `str` cannot be known at compilation time
 --> src/lib.rs:6:22
````
So although this trait fits object safety rules, i still can't implement it for unsized types.

**Comment by crlf0710** [comment]
Yes, i think this is a reasonable workaround. But it will be nice to get this limitation removed

**Comment by KrishnaSannasi** [comment]
Yes, I agree, this is an annoying limitation.

**Comment by QuineDot** [comment]
[The original example can be worked around](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8024cb31819400a9a46daabed4ebdcb2) with the [trivial bounds RFC](https://github.com/rust-lang/rust/issues/48214) (you still cannot call the implementation).

That said, I still feel that the approximation of using `Sized` instead of some `NotDyn` trait should be fixed; said trait could be a supertrait of `Sized` for backwards compatibility.  The reason (well, beyond semantics) being, sometimes you _do_ want to call some generic method on a `str` implementation (imagine the method was `-> ()` in the example).

**Comment by kupiakos** [comment]
For reference, you can also work around this on stable with an unused [HRTB][hrtb]:
```rust
trait Foo {
    fn bar(&self) -> Self where Self: Sized;
}

impl Foo for str {
    fn bar(&self) -> Self where for<'a> str: Sized { unimplemented!() }
}
```

[hrtb]: https://doc.rust-lang.org/nomicon/hrtb.html

**Comment by cyrgani** [comment]
This was fixed by #135480, meaning that 
```rust
trait Foo {
    fn bar(&self) -> Self where Self: Sized;
}

impl Foo for str {}
```
compiles sucessfully now.
